### PR TITLE
Multiple small fixes in the Task and Task.Supervisor docs.

### PR DESCRIPTION
**Pull request by alco**

### CODE REVIEW

File: lib/elixir/lib/task.ex [hunk 1]
```diff
@@ -526,15 +532,15 @@ defmodule Task do
...
 
   It is not required to call this function when terminating the caller, unless
-  exiting with reason `:normal` or the task is trapping exits. If the caller is
-  exiting with a reason other than `:normal` and the task is not trapping exits the
+  exiting with reason `:normal` or if the task is trapping exits. If the caller is
+  exiting with a reason other than `:normal` and the task is not trapping exits, the
   caller's exit signal will stop the task. The caller can exit with reason
-  `:shutdown` to shutdown linked processes, such as tasks, that are not trapping
-  exits without generating any log messages.
+  `:shutdown` to shutdown all of its linked processes, including tasks, that
```
Comment (alco) [inline lib/elixir/lib/task.ex]: Is this correct or should I change it back?

File: lib/elixir/lib/task.ex [hunk 2]
```diff
@@ -526,15 +532,15 @@ defmodule Task do
   The shutdown method is either a timeout or `:brutal_kill`. In case
   of a `timeout`, a `:shutdown` exit signal is sent to the task process
   and if it does not exit within the timeout it is killed. With `:brutal_kill`
-  the task is killed straight away. In case the task exits abnormally, or a 
+  the task is killed straight away. In case the task exits abnormally, or a
   timeout shutdown kills the task, this function will exit with the same reason.
```
Comment (alco) [inline lib/elixir/lib/task.ex]: I don't understand what "a timeout shutdown kills the task" means.

Comment (fishcakez) [inline lib/elixir/lib/task.ex]: It is ok to remove that bit. It means if the task is shutdown or killed by another process.

File: lib/elixir/lib/task.ex (hunk 1 above)
Comment (fishcakez) [inline lib/elixir/lib/task.ex]: This change is fine, the meaning is the same unless I missed something?

File: lib/elixir/lib/task.ex [hunk 3]
```diff
@@ -344,7 +344,8 @@ defmodule Task do
   This function returns a tuple with the returned value
   in case the message matches a task that exited with
   success alongside the matching task. It returns `nil`
-  if no task was found. It exits if the task has failed.
+  if no task was found. It exits if any of the tasks have
+  failed.
```
Comment (fishcakez) [inline lib/elixir/lib/task.ex]: This change is misleading. The return or exit is based purely on the message and so relates to "the" task. Another task could have failed but it won't cause an exit.

File: lib/elixir/lib/task.ex (hunk 2 above)
Comment (fishcakez) [inline lib/elixir/lib/task.ex]: E.G. If a supervisor is terminating the task with a timeout shutdown strategy and the timeout is exceeded it will kill the task.

File: lib/elixir/lib/task.ex [hunk 4]
```diff
@@ -12,46 +12,46 @@ defmodule Task do
       res  = do_some_other_work()
       res + Task.await(task)
 
-  Tasks spawned with `async` can be waited on by their caller
+  Tasks spawned with `async` can be awaited on by their caller
```
Comment (alco) [inline lib/elixir/lib/task.ex]: There was a commit that changed the `awaited` to `waited` here, but I changed it back so that the same word is everywhere in the module.

File: lib/elixir/lib/task.ex (hunk 1 above)
Comment (alco) [inline lib/elixir/lib/task.ex]: I just wasn't sure whether it's correct to generalize the behaviour to "all linked processes". Didn't find any special treatment of the `shutdown` exit reason in the Erlang docs.

File: lib/elixir/lib/task.ex (hunk 2 above)
Comment (alco) [inline lib/elixir/lib/task.ex]: Thanks, I have rephrased it.

File: lib/elixir/lib/task.ex (hunk 3 above)
Comment (alco) [inline lib/elixir/lib/task.ex]: My bad, I misunderstood the explanation. I tried to make it clearer in a separate commit. Thanks for catching this.

**Comment by alco** [comment]
Added new commits so that it's easier to see what changed in response to your feedback. I can squash it into a single commit once it's been reviewed.

File: lib/elixir/lib/task.ex [hunk 5]
```diff
@@ -341,10 +339,12 @@ defmodule Task do
...
   a task that matches the given message.
 
-  This function returns a tuple with the returned value
-  in case the message matches a task that exited with
-  success alongside the matching task. It returns `nil`
-  if no task was found. It exits if the task has failed.
+  If the message matches a task that finished normally,
+  this function returns a tuple with the task's
+  result and `Task` struct. If the message matches a
+  task that terminated abnormally, this function will exit.
```
Comment (fishcakez) [inline lib/elixir/lib/task.ex]: It need not be abnormally.
```elixir
Task.async(fn() -> exit(:normal) end)
```
Will cause the task to exit normally and cause an exit. It should be fine just to remove "abnormally".

Comment (alco) [inline lib/elixir/lib/task.ex]: OK, so I had to read the source code to finally understand what this function is doing. What do you think about the following wording:

```
  If the message matches a task that finished normally,
  this function returns a tuple with the task's
  result and `Task` struct. If the message matches a
  task that terminated before sending its result back,
  this function will exit.
```

I didn't want to say just "terminated" because that is more likely to raise question about the exact semantics.

As a slight offtopic, the exiting behaviour of this function surprises me. If the common use case for `find` is to check whether an incoming message is related to a task, I would expect to get an error tuple back for a prematurely terminated task. I guess this current behaviour is supposed to mirror what happens when you `await`?

/cc @josevalim

Comment (josevalim) [inline lib/elixir/lib/task.ex]: Yes, it was meant to mirror await. Plus we cannot change it because it
would be backwards incompatible.

On Wednesday, December 30, 2015, Alexei Sholik <notifications@github.com>
wrote:

> In lib/elixir/lib/task.ex
> <https://github.com/elixir-lang/elixir/pull/4128#discussion_r48607917>:
>
> > @@ -341,10 +339,12 @@ defmodule Task do
> >    Receives a group of tasks and a message and finds
> >    a task that matches the given message.
> >
> > -  This function returns a tuple with the returned value
> > -  in case the message matches a task that exited with
> > -  success alongside the matching task. It returns `nil`
> > -  if no task was found. It exits if the task has failed.
> > +  If the message matches a task that finished normally,
> > +  this function returns a tuple with the task's
> > +  result and `Task` struct. If the message matches a
> > +  task that terminated abnormally, this function will exit.
>
> OK, so I had to read the source code to finally understand what this
> function is doing. What do you think about the following wording:
>
>   If the message matches a task that finished normally,
>   this function returns a tuple with the task's
>   result and `Task` struct. If the message matches a
>   task that terminated before sending its result back,
>   this function will exit.
>
> I didn't want to say just "terminated" because that is more likely to
> raise question about the exact semantics.
>
> As a slight offtopic, the exiting behaviour of this function surprises me.
> If the common use case for find is to check whether an incoming message
> is related to a task, I would expect to get an error tuple back for a
> prematurely terminated task. I guess this current behaviour is supposed to
> mirror what happens when you await?
>
> /cc @josevalim <https://github.com/josevalim>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4128/files#r48607917>.
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

File: lib/elixir/lib/task.ex [hunk 6]
```diff
@@ -12,46 +12,46 @@ defmodule Task do
...
   into concurrent code with `Task.async/1` while keeping its semantics.
   When invoked, a new process will be created, linked and monitored
   by the caller. Once the task action finishes, a message will be sent
   to the caller with the result.
 
   `Task.await/2` is used to read the message sent by the task.
-  `await` will check the monitor setup by the call to `async/1` to
+  `await` will check the monitor set up by the call to `async/1` to
   verify if the process exited for any abnormal reason (or in case
   exits are being trapped by the caller).
```
Comment (alco) [inline lib/elixir/lib/task.ex]: I don't think this paragraph needs to say anything else after the first sentence. Talking about checking the monitor implies that the caller will not be forcibly terminated if the task exits, which is not true.

And also this is somewhat weird behaviour: if the caller traps exits, it presumably doesn't want to be killed. But then it calls `Task.await` which calls `exit()` if the task has crashed. The moduledoc already advises against trapping exits. I would add a sentence to it explaining that trapping exits does not prevent the task owner dying when it calls `Task.await`.

Comment (alco) [inline lib/elixir/lib/task.ex]: Correction: the advise against trapping exits is mentioned in the doc for `async`.

File: lib/elixir/lib/task.ex [hunk 7]
```diff
@@ -338,13 +341,17 @@ defmodule Task do
...
-  This function returns a tuple with the returned value
-  in case the message matches a task that exited with
-  success alongside the matching task. It returns `nil`
-  if no task was found. It exits if the task has failed.
+  If the message matches a task that finished normally,
+  this function returns a tuple with the task's
+  result and `Task` struct.
+
+  If the message matches a task that terminated before
+  sending its result back, this function will exit.
```
Comment (alco) [inline lib/elixir/lib/task.ex]: I still find this behaviour somewhat unsettling. 

When starting a task outside of an OTP behaviour, I can use `Task.Supervisor.async_nolink` and then call `Task.yield` and get back a nice `{:exit, reason}` for a failed task.

However, when spawning tasks inside a gen server, even using `async_nolink`, the user seems to have no choice but to exit when they call `find`.

**Comment by alco** [comment]
I did another pass over the whole module (will clean up the history before merging). Please take another look.

File: lib/elixir/lib/task.ex (hunk 7 above)
Comment (alco) [inline lib/elixir/lib/task.ex]: So yeah, it appears I'm still not 100% sold on the idea that task errors should propagate to the caller. Sorry, @fishcakez! I was 99% of the way there, but then I found out `yield` had been changed to not always exit in Elixir 1.2.

So I guess that choice between the exiting `await` and a non-exiting `yield` confuses me when I'm thinking of the right way to use tasks. And it's especially confusing in the case of `find` (which is supposed to be used in a gen server) since killing a gen server is something you don't normally expect. Normally, when someone is talking about a gen server spawning tasks, they imply receiving results from those tasks as messages and not killing the whole server because of one of the tasks terminating.

Comment (josevalim) [inline lib/elixir/lib/task.ex]: When `Task.find` was introduced, we didn't have `async_nolink` and changing it would be backwards incompatible. We were also not sure if the `Task.find` would stick. Today, there are two ways of using tasks:

1. You want it to compute something in asynchronously and you want it to go until the end (async + await)

2. You want to start some computation that you may not necessarily want to wait until it finishes (async + yield + shutdown)

At this point, I am actually wondering if we shouldn't just kill `Task.find`. If you have a server, you can easily use `Task.start_link` and send a message back. Today, the task message format is documented and unlikely to change, so you can also check for it explicitly.

Thoughts? Of course, it would have to be on 1.3.

Comment (josevalim) [inline lib/elixir/lib/task.ex]: Gone in master.

Comment (alco) [inline lib/elixir/lib/task.ex]: I agree with the two ways of using tasks you describe, those are the essential building blocks for performing async tasks concurrently.

Neither of them fits well into a gen server though. I'm not proposing to remove `Task.find` but I take it that you share my concerns about its utility. It almost seems to me that it should have behaved like `Task.yield` and returned `{:exit, reason}` instead of exiting. Then it would have been able to work nicely together with `async_nolink`. But let's just finish this PR and defer any further design discussion until another time.

My last question would then be: should we leave the `## Compatibility with OTP behaviours` section I added above? Perhaps, in the doc for `find/2` I should reiterate the point about adjusting the task code to return a tagged tuple from `async/3`'s doc:

> If you don't want the task to fail then you must change the `heavy_fun/0` code in the same way you would achieve it if you didn't have the async call. For example, to either return `{:ok, val} | :error` results or, in more extreme cases, by using `try/rescue`.

I don't @fishcakez would be happy about people sticking `try/rescue` in their tasks, but, in my opinion, this would serve as a workaround approximating the desired behaviour of `Task.find` I described in the second paragraph of this comment.

Regardless, I'm willing to just remove the `OTP Compatibility` section above and merge this if you don't have any further comments about my changes.

Comment (alco) [inline lib/elixir/lib/task.ex]: For the record, I only just now saw @josevalim's comment

> Gone in master.

after I posted mine.

Comment (josevalim) [inline lib/elixir/lib/task.ex]: Your comment is relevant regardless of my comment. :)

Comment (fishcakez) [inline lib/elixir/lib/task.ex]: > Gone in master.

From 1.2 we can pattern match on maps with variable keys. Using a map of the form `%{reference() => any()}` will be much more efficient and likely cleaner than using `Task.find`. Particularly when using `async` because do not need to handle `:DOWN`s from tasks (unless trapping exits or calling `exit(:normal)` in the task).

> I don't think @fishcakez would be happy about people sticking try/rescue in their tasks, but, in my opinion, this would serve as a workaround approximating the desired behaviour of Task.find I described in the second paragraph of this comment.

I have suggested this approach on IRC a few times. When using a Task you should use exactly the same error handling as when not using a Task. Task is a method of providing concurrency, not error isolation.

I think this comes down to error handling in general. If a specific error is anticipated then handle (catch/rescue) it as close to the source as possible. If all errors are to be handled then the catch/trap point needs to be carefully thought out if its not going to be a supervisor.

> Normally, when someone is talking about a gen server spawning tasks, they imply receiving results from those tasks as messages and not killing the whole server because of one of the tasks terminating.

Is this true?

**Comment by josevalim** [comment]
Ping.

**Comment by alco** [comment]
I have rebased this on master and cleaned up the commit history. I've also updated the section `Compatibility with OTP behaviours` [here](https://github.com/alco/elixir/commit/4da7b4c11edbc7410a9610b4d26cf43a11beb99f) to read

> It is not recommended to `await` a long-running task inside an OTP behaviour such as `GenServer`. Instead, you should match on the message coming from a task inside your `handle_info` callback.

The format of the result message is described in `async/3`. Should we also mention the fact that a `:DOWN` message will arrive in `async_nolink`'s docstring?

**Comment by josevalim** [comment]
> The format of the result message is described in async/3. Should we also mention the fact that a :DOWN message will arrive in async_nolink's docstring?

Yes, we should say it is always monitored, so both regular and DOWN messages will still arrive.

**Comment by alco** [comment]
Updated.

**Comment by josevalim** [comment]
@fishcakez can you please give it one last pass and merge it? Thank you both!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by alco** [comment]
@josevalim @fishcakez ping? :)

File: lib/elixir/lib/task.ex [hunk 8]
```diff
@@ -359,22 +365,24 @@ defmodule Task do
...
 
   @doc """
-  Yields for a task reply in the given time interval.
+  Temporarily blocks the current process waiting for a task reply.
 
-  Returns `{:ok, reply}` if the reply is received, `{:exit, reason}`
-  if the task exited or `nil` if no reply arrived.
+  Returns `{:ok, reply}` if the reply is received, `nil` if
+  no reply has arrived, or `{:exit, reason}` if the task exited.
+  Keep in mind that normally this function will exit if the task
```
Comment (fishcakez) [inline lib/elixir/lib/task.ex]: "this function" won't exit, rather its an exit signal.

File: lib/elixir/lib/task.ex [hunk 9]
```diff
@@ -359,22 +365,24 @@ defmodule Task do
...
+  process terminates abnormally. `{:exit, reason}` can only be
+  returned if the task process exited with the reason `:normal`
+  or if it isn't linked to the caller.
```
Comment (fishcakez) [inline lib/elixir/lib/task.ex]: Trapping exits?

**Comment by fishcakez** [comment]
There is just one paragraph which is a bit confusing. Perhaps you could reword that? Sorry for dropping the ball on the review @alco.

File: lib/elixir/lib/task.ex (hunk 4 above)
Comment (heavyk) [inline lib/elixir/lib/task.ex]: is the correct english is "awaited on by their *calling* process (and only *the process that called it*)"?

at least it sounds better to my ears, but maybe it's called the 'caller' somewhere else.

File: lib/elixir/lib/task.ex [hunk 10]
```diff
@@ -133,7 +131,7 @@ defmodule Task do
   works with anonymous functions. That's because anonymous functions expect
   the same module version to exist on all involved nodes. Check the `Agent` module
   documentation for more information on distributed processes as the limitations
-  described in the agents documentation apply to the whole ecosystem.
+  described there apply to the whole ecosystem.
```
Comment (heavyk) [inline lib/elixir/lib/task.ex]: I think you're missing a comma after there

Comment (alco) [inline lib/elixir/lib/task.ex]: I don't see why it needs a comma. Could you explain?

**Comment by alco** [comment]
@fishcakez Addressed your two latest comments.

File: lib/elixir/lib/task.ex (hunk 4 above)
Comment (alco) [inline lib/elixir/lib/task.ex]: This is a valid point. As of the this PR, the docs currently use the terms `caller`, `calling process`, `owner`, and `owning process` somewhat interchangeably. @josevalim Should something be done about it?

File: lib/elixir/lib/task.ex (hunk 10 above)
Comment (heavyk) [inline lib/elixir/lib/task.ex]: actually looking again now, I don't know if it's even right or very necessary, but it would better go after distributed processes.

"Check the `Agent` module documentation for more information on distributed processes, as the limitations described there apply to the whole ecosystem."
